lte/gateway/c/session_manager/SessionStateEnforcer.cpp

```diff
@@ -0,0 +1,245 @@
+#include <time.h>
+#include <utility>
+#include <vector>
+
+#include <google/protobuf/repeated_field.h>
+#include <google/protobuf/timestamp.pb.h>
+#include <google/protobuf/util/time_util.h>
+#include <grpcpp/channel.h>
+#include "magma_logging.h"
+#include "EnumToString.h"
+#include "SessionStateEnforcer.h"
+
+namespace magma {
+// temp routine
+void call_back_void_upf(grpc::Status, magma::UpfRes response) {
+  // do nothinf but to only passing call back
+}
+
+/*constructor*/
+SessionStateEnforcer::SessionStateEnforcer(
+    std::shared_ptr<StaticRuleStore> rule_store, SessionStore& session_store,
+    std::shared_ptr<PipelinedClient> pipelined_client,
+    std::shared_ptr<AmfServiceClient> amf_srv_client,
+    magma::mconfig::SessionD mconfig)
+    : session_store_(session_store),
+      pipelined_client_(pipelined_client),
+      amf_srv_client_(amf_srv_client),
+      retry_timeout_(1),
+      mconfig_(mconfig)
+{
+  // for now this is the right place, need to move if find  anohter right place
+  static_rule_init();
+}
+
+void SessionStateEnforcer::attachEventBase(folly::EventBase* evb) {
+  evb_ = evb;
+}
+
+void SessionStateEnforcer::stop() {
+  evb_->terminateLoopSoon();
+}
+
+folly::EventBase& SessionStateEnforcer::get_event_base() {
+  return *evb_;
+}
+
+bool SessionStateEnforcer::m5g_init_session_credit(
+    SessionMap& session_map, const std::string& imsi,
+    const std::string& session_id, const SessionConfig& cfg) {
+  /* creating SessionState object with state CREATING
+   * This calls constructor and allocates memory*/
+  auto session_state =
+      std::make_unique<SessionState>(imsi, session_id, cfg, *rule_store_);
+  MLOG(MINFO) << " New SessionState object created with IMSI: " << imsi
+              << " session context id : " << session_id;
+  handle_session_init_rule_updates(session_map, imsi, *session_state);
+
+  /* Find same UE or imsi already present, if not add
+   * TODO - Need to check if same DNN/APN already exist
+   */
+  auto exist_imsi = session_map.find(imsi);
+  if (exist_imsi == session_map.end()) {
+    // First time a session is created for IMSI in the SessionMap
+    MLOG(MDEBUG) << "First session for IMSI " << imsi
+                 << " with session context ID " << session_id;
+    session_map[imsi] = std::vector<std::unique_ptr<SessionState>>();
+  } else {
+    MLOG(MINFO) << "Sanjay inside else session exist session for IMSI " << imsi;
+    session_map[imsi].push_back(std::move(session_state));
+  }
+  return true;
+}
+
+void SessionStateEnforcer::handle_session_init_rule_updates(
+    SessionMap& session_map, const std::string& imsi,
+    SessionState& session_state) {
+  auto itp = pdr_map_.equal_range(imsi);
+  for (auto itr = itp.first; itr != itp.second; itr++) {
+    /* Get the PDR numbers, now  get the rules from global static rule
+     * list
+     */
+    SetGroupPDR rule;
+    GlobalRuleList.get_rule(itr->second, &rule);
+    session_state.insert_with_static_rules(&rule);

```

Comment:
I am not sure if I am missing something, but is it possible to look for a better naming
`insert_static_rule`?

Reply:
yes changed it to "insert_pdr